### ISSUE TITLE
opening and saving a file in one pipe-command not possible?

### ISSUE THREAD

**Issue by Herr-Sepp** [PowerShell/PowerShell#14925]
## Steps to reproduce

```powershell
@"
I
like
Powershell
"@ > .\file.txt

1) Get-Content .\file.txt | Where Length -eq 10 | Set-Content .\file.txt
   Get-Content .\file.txt | Where-Object { $_.Length -eq 10 } | Set-Content .\file.txt

2) Get-Content .\file.txt | Where Length -eq 10 | Out-File .\file.txt

3) Get-Content .\file.txt | Where Length -eq 10 | Out-File -Append .\file.txt


```

## Expected behavior

```none
1 & 2) file.txt with "Powershell" as content
3) "Powershell" attached to file.txt
```

## Actual behavior

```none
1) Powershell Error: Set-Content: The process cannot access the file 'C:\file.txt' because it is being used by another process.
2) File is empty
3) Endless loop

This is really not what I expected, especially dangerous as sometimes it looks like it works

Get-Content .\file.txt | Where $true | Set-Content .\file.txt    ✅
Get-Content .\file.txt | Where {$true} | Set-Content .\file.txt  ❌
Get-Content .\file.txt | Where Length -eq 3 | Set-Content .\file.txt ⁉ No error but file is empty
```

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.1.2
PSEdition                      Core
GitCommitId                    7.1.2
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by p0W3RH311** [comment]
you needs parhentesis:

```
PS> @"
>> I
>> like
>> Powershell
>> "@ > .\file.txt
PS> ( Get-Content .\file.txt | Where Length -eq 10) | Set-Content .\file.txt
PS> Get-Content .\file.txt
Powershell

```

**Comment by Herr-Sepp** [comment]
@p0W3RH311 wow thanks a lot.
Can you maybe also explain why the parhentesis are necessary here?
I have a hard time explaining the behavior without them.

PS: do you think I should close this ticket because it is an expected behavior of Powershell?

**Comment by p0W3RH311** [comment]
@Herr-Sepp 

in pipeline view:
```

0) get-content ... | where ... | set-content ...
1) PROCESS LINE1 and wait | PROCESS LINE 1 and wait | PROCESS {TRY TO WRITE to a read open file}
```
# in windows unlike unix its not permited to write in open file

```

0) (get-content ... | where ...) | set-content ...
1) (PROCESS { SEND LINE 1 & wait } | PROCESS { FILTER LINE 1 & wait }) 
2) (PROCESS { SEND LINE 2 & wait } | PROCESS { FILTER LINE 2 & wait }) 
...
...
3) (END {} | END {})  | PROCESS { THE FILTERED RESULT} END {} 

```

**Comment by mklement0** [comment]
@Herr-Sepp, if I may weigh in and add to @p0W3RH311's comments:

Yes, I think this issue can be closed, because the behavior is by design.

The parentheses are necessary, because they force the file to be _read in full_ and _closed_ again _before_ the later pipeline segments attempt to open the output file for writing, which _truncates_ the file (discards its existing content).

---

**tl;dr**:

**If you want to write back to an input file _in the same pipeline_:**

* ALWAYS use `(...)` to enclose the _reading_ command, knowing that:
  * The file's (possibly transformed) content must then fit into memory _as a whole_, because the file's _entire content_ is read and the file is closed _before_ the output from the `(...)` expression is processed further.
  * There's a hypothetical risk of data loss, if the process of writing back to the original file gets interrupted (such as a power loss at the exact moment the file is being written).
    * The safe - but much more cumbersome - alternative is to write to a _temporary_ file first, and, on successfully completing writing to it, replace the original file with the temporary one.


* ALWAYS use a _separate pipeline segment_ to perform the _writing_ part.
  * NEVER use `>` _as part of the same command / pipeline segment_ - complete data loss will occur, even with `(...)`
  * NEVER use `>>` _as part of the same command / pipeline segment_ - an infinite loop with an ever-growing file will result. 


---

**The pitfall of accidentally erasing the content of an input file you're trying to write back to in the same pipeline is not specific to PowerShell.**

The simplest example, which shows the problem in both PowerShell and POSIX-compatible shells is:

```sh
# Note: On Windows, `cat` is an alias of `Get-Content`, but not on Unix.
echo hi > file; cat file > file; cat file  # -> !! no output, because file's content was discarded by `> file`
```

This **quietly erases the content of file `file`, leaving it with a size of `0`**, because the `>` redirection operator (effectively an alias of `Out-File` in PowerShell) opens `file` for (non-appending) writing and thereby truncates it _before_ `cat` gets a chance to read it.

Therefore, **if you use`>` to write to the file in the _same pipeline segment_, data loss occurs even if you _do_ enclose the reading (`Get-Content`) command in `(...)`**:

```powershell
# !! DATA LOSS: Redirection operator `>` must not be used in same pipeline (segment) as (Get-Content ...)
PS> echo hi > file; (Get-Content file) > file; (Get-Item file).Length
0 # !! file was truncated before Get-Content got a chance to read it.
```

**The `(...)` technique _only_ works if writing back to the file occurs in a _later pipeline segment_**, when you pipe to cmdlets such as `Out-File`, `Set-Content`, or `Export-Csv`.

```powershell
# OK: (Get-Content ...) (parentheses required) reads everything up front and closes the file, so that Set-Content in 
#     a *later pipeline segment* can write back to it.
PS> echo hi > file; (Get-Content file) | Set-Content file; Get-Content file
hi #  OK
```
---

**The following discusses platform differences with respect to file locking, the read buffering that occurs behind the scenes, and a `Set-Content` implementation detail**: 

_Without_ `(...)`, the file's content is _typically_ erased even if you use the file-writing command in a _later_ pipeline segment, but `Set-Content` is an _exception_, albeit one that cannot be relied upon:

By a _lucky accident_ falling out from a - debatable - `Set-Content` implementation detail, data loss _on Windows only_ is _prevented_ , whereas on _Unix_ possibly initially undetected _partial data loss is likely_ :

* Unlike other cmdlets, `Set-Content` apparently defers opening the output file for writing until its `ProcessRecord()` method is called (whereas that normally happens in `BeginProcessing()`), i.e. only _when the first input object is received_.

  * As an aside: This means that `@() | Set-Content file` neither creates a previously nonexistent file `file`  nor does it overwrite a preexisting one.

* This means that the input-providing `Get-Content` (except if enclosed in `(...)`) already has an open file handle for the file in question at the time `Set-Content` tries to open it for writing, and what happens then depends on the host platform:

  * On _Windows_, the open-for-writing attempt _fails_, because the file is considered _locked_:
    *  The error you saw is emitted, the file is _not_ modified  and therefore _no data loss occurs_:

       * `Set-Content: The process cannot access the file '...' because it is being used by another process.`

  * On _Unix_-like platforms,  the open-for-writing attempt _succeeds_, because the file is _not_ considered _locked_ for writing, but _partial data loss is likely_:

     *  The attempt to rewrite the file succeeds _only if the file happens to be smaller than the buffer size used to read from the file_, which I presume is an implementation detail of the underlying .NET APIs, and is currently `4 KB` (`4096 bytes`).

     * The reason is that one buffer's worth of data has already been read into memory by the time `Set-Content` truncates the file, so that one buffer's worth data is passed through, but the rest of the input data is _lost_ at that point, as the following example demonstrates
         * `Set-Content file -Value (, 'x' * 512kb); Get-Content file | Set-Content file; if ($?) { (gi file).length / 1kb }`
         * The above creates a `512mb` file, and then attempts to use `Get-Content | Set-Content` to rewrite it in place. Because the file gets truncated after one buffer's worth has been read, only the first `4096` bytes of the original content ends up in the file.

     * The only _exceptions_ where larger-than-the-buffer-size files _are_ correctly processed without `(...)` are:

         * If you use `-Raw` or `-ReadCount 0`, because then all input has by definition already been read.

           *  Note: Unlike on Windows, the fact that the file handle apparently hasn't yet been closed at that point is _not_ a problem on Unix; on Windows, even use of `-Raw` / `-ReadCount 0` requires enclosing in `(...)`; in short: in cross-platform code you need to use `(...)` anyway.

         * If a given file happens to contain only _one_ (very long) line, because PowerShell keeps reading until at least one line has been read in full, even beyond the buffer size, which in this particular case again means that the file is read _in full_ first.
